### PR TITLE
WP-1085 Visual testing harness

### PR DESCRIPTION
**Pull request by sergeychernyshev**
Visual test harness that includes some jest & babel configuration, `testServer.js` and some test boilerplate in `Button.test.jsx`

It lets us create automated tests for visual discrepancy which is critical for projects which concentrate on styles rather than functionality.

JIRA: https://meetup.atlassian.net/browse/WP-1085

### CODE REVIEW

**Comment by hsbacot** [comment]
How would we feel about altering the interface here and replacing `createRoute(slug, element)` with something like `renderToPage(element)` or `renderToPage(server, element)`?

By passing the server we could handle the route creation within the function, but then it has side effects we need to make users aware of.

This would allow us to skip managing a slug in the tests since we don't really need to know the url of the test server as long as it is consistent with the one returned by `createRoute` or `renderToPage`.

Thoughts @sergeychernyshev @mmcgahan?

**Comment by mmcgahan** [comment]
@hsbacot I like the recommendation to make the user-facing rendering interface simpler - it would be nice to remove any reference to 'routes' (and potentially to `server` and `puppeteer`) since those are implementation details.

If it was me, I'd have some fun writing a wrapper that provides a `renderer` that maybe needs to be started and stopped, but otherwise just takes an element and returns a screenshot async. The public interface could work something like this:

```jsx
import screenRenderer from './screenRenderer';

beforeAll(() => {
  renderer = screenRenderer(config);
}
afterAll(() => {
  renderer.stop();
}

test('foo', async () => {
  expect(await renderer.screenshot(<Foo />)).toMatchImageSnapshot();
});
```

However, that might lean into over-engineering at this point... @sergeychernyshev if you feel like you have time for something like that, which could be used as a general utility in any of our codebases, that might be a nice thing to do, or we can leave it for followup once we confirm that this screenshot testing is serving the purpose we need it to.

**Comment by sergeychernyshev** [comment]
The only reason SLUG is external is because I tried to match the filename to the test name more or less closely, but if we go with the default `toMatchImageSnapshot()` file naming convention, then we can easily hide it this way.

There is a challenge in making this more generic is that styles are hard-coded and decision to render HTML on server without rehydrating is pretty strict for a generic use-case, but it can be generalized later as well.

So, "my time is your time" and I can do the basic abstraction at the expense of other tickets which I don't see the abundance of in our current sprint anyway ;)

**Comment by sergeychernyshev** [comment]
OK, update is in, please review again, @mmcgahan @hsbacot

**Comment by sadafie** [comment]
I just went over this PR with Sergey during our pairing. One question that came to mind is the behavior of the boilerplate server + browser initialization and teardown. The way it's written at the moment, each test suite will:

- Create and initialize a hapi server
- Start up a new browser instance
- Create an html page per test + point the browser to it
- Kill the server (and browser?)

If we end up with a test file/suite per component, then we'd be doing this setup and teardown N times for N components. Right now this is relatively quick for one test suite, but how can we optimize this or should we even?

File: packages/swarm-components/src/Button.test.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+describe('👀 components are visually the same', function() {
+	let renderer;
+
+	// This is ran when the suite starts up.
+	beforeAll(async () => {
+		renderer = await getScreenRenderer({
+			port: 4000,
+			viewport: { width: 200, height: 100 },
+			staticPath: '../../swarm-styles/dist',
```
Comment (sadafie) [inline packages/swarm-components/src/Button.test.jsx]: The way this staticPath is defined is strange to me. This is nice for testing in development (i.e., you make a change in swarm-styles and run `yarn build` and you'll be referencing the latest styles here), however, from a proper dependency management viewpoint I feel as if `swarm-styles` should be a devDependency of `swarm-components` and we reference `main.css` and `global.css` in fashion similar to how we do it in mup-web currently.

Comment (sergeychernyshev) [reply packages/swarm-components/src/Button.test.jsx]: Agreed. In fact TravisCI might choke as it build builds packages independently.

Comment (sadafie) [reply packages/swarm-components/src/Button.test.jsx]: One edge case scenario I can imagine this breaking in - 

You have changes in `swarm-styles` and you run these tests against those changes. You push up the changes to these tests, however for some reason you don't push up changes to `swarm-styles`. Ideally tests will fail and be caught in Travis, but you need to know what you're looking for.

**Comment by hsbacot** [comment]
@sadafie good thoughts on the server instances, I feel like that is beyond how the harness is written but should be a consideration when implementing. I would imagine that each suite would reference the same server instance. So in the component tests, we'd have a setup file that allows for this.

**Comment by hsbacot** [comment]
Jest global setup and tearDown docs -> https://jestjs.io/docs/en/configuration.html#globalsetup-string

**Comment by sadafie** [comment]
@hsbacot Yep, cool. I just wanted us thinking about it. There's about a dozen or so tickets in the backlog (thank you @sergeychernyshev!) for writing these tests. Perhaps after this branch merges/before the next sprint planning, @sergeychernyshev or someone else can take one of those tickets and build out the setup step so then multiple folks can easily contribute to all the tests.

File: packages/swarm-components/package.json [hunk 2]
```diff
@@ -1,53 +1,60 @@
...
-    "babel-plugin-react-docgen": "2.0.2",
-    "flow-bin": "0.96.0"
-  },
-  "dependencies": {
-    "@meetup/swarm-constants": "^0.2.9",
-    "classnames": "2.2.6",
-    "html-loader": "0.5.5",
-    "swarm-icons": "3.8.354"
-  }
+	"name": "@meetup/swarm-components",
```
Comment (mmcgahan) [inline packages/swarm-components/package.json]: I know it's a little annoying, but `.json` files should use spaces instead of tabs, especially `package.json`, which won't work with tabs in npm before version 5 https://github.com/prettier/prettier/issues/3533

Not sure if this was a Prettier change, but we should make sure json files get 2-space indentation

File: packages/swarm-components/src/screenRenderer.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	async stop() {
+		await this.browser.close();
+		return this.server.stop();
+	}
+
+	async screenshot(element) {
+		const page = await this.browser.newPage();
+		if (this.config.viewport) {
```
Comment (mmcgahan) [inline packages/swarm-components/src/screenRenderer.js]: feel free to ignore for now, but was wondering if the `screenshot` method should take an optional config override, at least for the viewport, so that elements with viewport-responsive styles can be rendered at different widths in tests.

Also, instead of testing for the existence of `this.config.viewport`, I think it would be better to define a default width and always call `page.setViewport`, since an 'undefined' viewport is ambiguous (presumably puppeteer has an internal default?)

e.g.

```js
async screenshot(element, shotConfig) {
  const page = await this.browser.newPage();
  page.setViewport(shotConfig && shotConfig.viewport || this.config.viewport);
  ...
}
```

File: packages/swarm-components/src/screenRenderer.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			page.setViewport(this.config.viewport);
+		}
+
+		const SLUG = `route-${this.routeIndex++}`;
```
Comment (mmcgahan) [inline packages/swarm-components/src/screenRenderer.js]: since this isn't a global constant, I recommend avoiding ALL_CAPS name - same with TEST_URL.

File: packages/swarm-components/src/screenRenderer.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		await page.goto(TEST_URL);
+
+		return page.screenshot();
+	}
+}
+
+const getScreenRenderer = async config => {
+	const renderer = new screenRenderer(config);
+
+	await renderer.init();
```
Comment (mmcgahan) [inline packages/swarm-components/src/screenRenderer.js]: if `init` had a `return this` at the end, `getScreenRenderer` could just be

```js
const getScreenRenderer = async config => new screenRenderer(config).init();
```

File: packages/swarm-components/package.json (hunk 2 above)
Comment (sergeychernyshev) [reply packages/swarm-components/package.json]: cc @hsbacot, looks like we keep changing tabs/spaces back and forth.

Do we have a standard for that? do we have a config we can just reuse?

File: packages/swarm-components/src/screenRenderer.js (hunk 3 above)
Comment (sergeychernyshev) [reply packages/swarm-components/src/screenRenderer.js]: Yes, puppeteer has internal default - we can standardize on something for this purpose or just explicitly use puppeteer's current default.

File: packages/swarm-components/package.json (hunk 2 above)
Comment (mmcgahan) [reply packages/swarm-components/package.json]: Looks like mup-web just solves this with a `.prettierignore` file https://github.com/meetup/mup-web/blob/master/.prettierignore

**Comment by sergeychernyshev** [comment]
> If we end up with a test file/suite per component, then we'd be doing this setup and teardown N times for N components. Right now this is relatively quick for one test suite, but how can we optimize this or should we even?

Did a quick run with 10 suits and 12 tests per suite, it took 77s on my laptop for initial run and 78-9s for re-runs (e.g. comparing to existing screenshots).

**Comment by mmcgahan** [comment]
sounds like it might be worth looking into optimizations as a followup if tests become slow. Particularly for PRs, we could potentially run Jest with something like [`--changedSince`](https://jestjs.io/docs/en/cli#changedsince) or [`--onlyChanged`](https://jestjs.io/docs/en/cli#onlychanged) to improve PR build time. Probably would still want to run the full test suite on merges to `master`

File: packages/swarm-components/package.json (hunk 2 above)
Comment (sergeychernyshev) [reply packages/swarm-components/package.json]: Hmm. It looks like this repo has plenty of things in `.prettierignore`: https://github.com/meetup/swarm-ui/blob/master/.prettierignore

@hsbacot, is this all intentional here?

**Comment by sadafie** [comment]
>Did a quick run with 10 suits and 12 tests per suite, it took 77s on my laptop for initial run and 78-9s for re-runs (e.g. comparing to existing screenshots).

@sergeychernyshev Did you do it using the shared setup/teardown that hunter referenced? https://github.com/meetup/swarm-ui/pull/46#issuecomment-482680256

If not, we should do that.

**Comment by sergeychernyshev** [comment]
I created a separate ticket to speed things up if they become problematic: https://meetup.atlassian.net/browse/WP-1113

File: packages/swarm-components/src/Button.test.jsx (hunk 1 above)
Comment (sergeychernyshev) [reply packages/swarm-components/src/Button.test.jsx]: @mmcgahan, @hsbacot - can you weight in on this.

Comment (mmcgahan) [reply packages/swarm-components/src/Button.test.jsx]: I think the thing being raised is that these tests are _integration_ tests of `swarm-styles + swarm-components`, not unit tests of `swarm-components`, which is how they are currently configured, so they need to be run in an environment where both packages are guaranteed to exist in their published form.

Two options I can think of:

1. Move these tests to a dedicated 'integration testing' package that depends on `swarm-components + swarm-styles` as dependencies, and imports their 'dist' modules when running tests. I think this would be the most 'formally correct' and robust solution, but it requires some planning to make sure that the publish + test workflow operates as expected
2. Make sure that the command that runs these integration tests is separate from the normal 'unit testing' command, and run the bundle/package command for `swarm-styles` as a prerequesite for running integration tests, e.g.
    ```
    "test:integration": "lerna run build --scope=@meetup/swarm-styles && jest --config jest.int.config.json"
    ```

Good call that these aren't strictly unit tests, though - I think it's easy to miss that.

Extending that thinking, however, is there a reason why component-specific styles are being managed in a package separate from the components? Why not keep tightly-coupled code in the same package? I can see why we would still want a separate `@meetup/swarm-styles` package for 'globals', but we probably want to consider re-organizing some of the component-specific styles to stay closer to their JSX counterparts long term. Separating packages by file type (CSS vs JS) isn't usually the best tactic

Comment (sergeychernyshev) [reply packages/swarm-components/src/Button.test.jsx]: I feel that we should go with option 2 for now, until we have a better feel for the whole testing setup and avoid premature optimization here.

I'll add a note to Travis CI ticket

Comment (sadafie) [reply packages/swarm-components/src/Button.test.jsx]: >Extending that thinking, however, is there a reason why component-specific styles are being managed in a package separate from the components? Why not keep tightly-coupled code in the same package? I can see why we would still want a separate @meetup/swarm-styles package for 'globals', but we probably want to consider re-organizing some of the component-specific styles to stay closer to their JSX counterparts long term. Separating packages by file type (CSS vs JS) isn't usually the best tactic

This @hsbacot would be best to comment on. In general I haven't had any issues with the current setup, up until this ticket where the distinction has made the setup somewhat awkward.

>Move these tests to a dedicated 'integration testing' package that depends on swarm-components + swarm-styles as dependencies, and imports their 'dist' modules when running tests. I think this would be the most 'formally correct' and robust solution, but it requires some planning to make sure that the publish + test workflow operates as expected

I'm in favor of option 1 as an immediate follow up especially if we plan to keep `swarm-components` and `swarm-styles` as separate packages for the time being. (i.e., I don't realistically see us wanting / prioritizing changing this).
